Firebase/Database/Core/FServerValues.m

```diff
@@ -20,37 +20,102 @@
 #import "FLeafNode.h"
 #import "FSnapshotUtilities.h"
 
+const NSString *kTimestamp = @"timestamp";
+const NSString *kIncrement = @"increment";
+
+@interface FServerValues ()
++ (id)resolveScalarServerOp:(NSString *)op
+           withServerValues:(NSDictionary *)serverValues;
++ (id)resolveComplexServerOp:(NSDictionary *)op
+                withExisting:(id<FNode>)existing
+                serverValues:(NSDictionary *)serverValues;
+@end
+
 @implementation FServerValues
 
 + (NSDictionary *)generateServerValues:(id<FClock>)clock {
     long long millis = (long long)([clock currentTime] * 1000);
-    return @{@"timestamp" : [NSNumber numberWithLongLong:millis]};
+    return @{kTimestamp : [NSNumber numberWithLongLong:millis]};
 }
 
 + (id)resolveDeferredValue:(id)val
-          withServerValues:(NSDictionary *)serverValues {
-    if ([val isKindOfClass:[NSDictionary class]]) {
-        NSDictionary *dict = val;
-        if (dict[kServerValueSubKey] != nil) {
-            NSString *serverValueType = [dict objectForKey:kServerValueSubKey];
-            if (serverValues[serverValueType] != nil) {
-                return [serverValues objectForKey:serverValueType];
-            } else {
-                // TODO: Throw unrecognizedServerValue error here
-            }
-        }
+              withExisting:(id<FNode>)existing
+              serverValues:(NSDictionary *)serverValues {
+    if (![val isKindOfClass:[NSDictionary class]]) {
+        return val;
+    }
+    NSDictionary *dict = val;
+    id op = dict[kServerValueSubKey];
+
+    if (op == nil) {
+        return val;
+    } else if ([op isKindOfClass:NSString.class]) {
+        return [FServerValues resolveScalarServerOp:op
+                                   withServerValues:serverValues];
+    } else if ([op isKindOfClass:NSDictionary.class]) {
+        return [FServerValues resolveComplexServerOp:op
+                                        withExisting:existing
+                                        serverValues:serverValues];
     }
     return val;
 }
 
++ (id)resolveScalarServerOp:(NSString *)op
+           withServerValues:(NSDictionary *)serverValues {
+    return serverValues[op];
+}
+
++ (id)resolveComplexServerOp:(NSDictionary *)op
+                withExisting:(id<FNode>)existing
+                serverValues:(NSDictionary *)serverValues {
+    // Only increment is supported as of now
+    if (op[kIncrement] == nil) {
+        return nil;
+    }
+
+    NSNumber *delta = op[kIncrement];
+    if (![existing isLeafNode]) {
+        return delta;
```

Comment:
If I am not mistaken (and I may very well be) "op[kIncrement]" could return something that is not an NSNumber. We might need to verify that it is indeed of a numeric type.

Reply:
It's probably fine to just add an assert.

Reply:
Actually this is an expected case. For example, a null value is not a leaf node and we want to set the value to delta in this case.